packages/babel-helper-member-expression-to-functions/src/index.js

```diff
@@ -100,6 +100,26 @@ const handle = {
       return;
     }
 
+    // { KEY: MEMBER } = OBJ -> { KEY: _ref } = OBJ; _set(MEMBER, _ref);
+    // [MEMBER] = ARR -> [_ref] = ARR; _set(MEMBER, _ref);
+    if (
+      (parentPath.isObjectProperty({ value: node }) &&
+        parentPath.parentPath.isObjectPattern()) ||
+      parentPath.isArrayPattern()
+    ) {
+      const { scope } = member;
+      const ref = scope.generateUidIdentifierBasedOnNode(node);
+      scope.push({ id: ref });
+
+      const assignmentParent = parentPath.getStatementParent();
+      assignmentParent.insertAfter(
```

Comment:
I'm not sure we can insert after the statement, because later patterns can default to the property we're assigning:

```js
[
  this.#foo,
  x = this.#foo
] = [0];
```

Should have `x === 0`. The only way I can see this working in 100% of cases is if we use a special setter property:

```js
// input
[this.#foo, x = this.#foo] = [0];

// output
const _this = this;
const internal = { set foo(v) { _this.#foo = v } };

[internal.foo, x = this.#foo] = [0]
```

Reply:
so the order matters for array pattern right?
eg:
```
[this.#foo, x = this.#foo] = [0]
```
vs
```
[x = this.#foo, this.#foo] = [0];
```

how about object destructuring?
```
{
   x = this.#foo,
   y: this.#foo,
   z = this.#foo,
} = { y: 1 }
```

Reply:
Yup, you're correct in both cases.

Reply:
would it be better if we defined `const internal = { set foo(v) { _this.#foo = v } };` at the constructor if we know there will be a destructuring, and reuse it throughout the scope of the class?

Reply:
Yah, I would only create the object when we hit a destructure.